core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1119,6 +1131,21 @@ class ReplicaManager(val config: KafkaConfig,
         else
           Set.empty[Partition]
 
+        // in all of the partitions that the current broker is now a follower of
+        // add all the topics in a list as candidate to remove metrics
+        val topicCandidates = mutable.Set[String]()
+        partitionsBecomeFollower.foreach(partition => topicCandidates.add(partition.topic))
+
+        // now iterate through all the partitions which the broker is a leader of
+        // and remove any topic that is found since the current broker still needs to update its metrics
+        // for those topics
+        partitionsBecomeLeader.foreach(partition =>
```

Comment:
Check against `partitionsBecomeLeader` would not be sufficient: there maybe some existing partitions whose leader does not change, whereas `partitionsBecomeLeader` only contains the partitions that leader would be migrated in.

Reply:
Would you think that `leaderPartitionsIterator` is more sufficient in this case?

Reply:
Yes, that should be sufficient. Note that this is defined as a function so each time you call it, this iterator is dynamically created from `nonOfflinePartitionsIterator` which is in turn created from `allPartitions`.